Flaky-test: AdminApi2Test.resetClusters

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/3232396920/jobs/5309016207

### Exception stacktrace

```
  Error:  Tests run: 77, Failures: 1, Errors: 0, Skipped: 61, Time elapsed: 478.051 s <<< FAILURE! - in org.apache.pulsar.broker.admin.AdminApi2Test
  Error:  resetClusters(org.apache.pulsar.broker.admin.AdminApi2Test)  Time elapsed: 0.235 s  <<< FAILURE!
  org.apache.pulsar.client.admin.PulsarAdminException$ServerSideErrorException: 
  
   --- An unexpected error occurred in the server ---
  
  Message: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  
  Stacktrace:
  
  org.apache.pulsar.metadata.api.MetadataStoreException: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.getException(ZKMetadataStore.java:449)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.handleDeleteResult(ZKMetadataStore.java:279)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.lambda$batchOperation$7(ZKMetadataStore.java:191)
  	at org.apache.zookeeper.MockZooKeeper.multi(MockZooKeeper.java:1006)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.batchOperation(ZKMetadataStore.java:165)
  	at org.apache.pulsar.metadata.impl.batching.AbstractBatchedMetadataStore.flush(AbstractBatchedMetadataStore.java:117)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
  	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
  	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  	at java.base/java.lang.Thread.run(Thread.java:833)
  Caused by: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  	at org.apache.zookeeper.KeeperException.create(KeeperException.java:132)
  	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.getException(ZKMetadataStore.java:439)
  	... 12 more
  
  	at org.apache.pulsar.client.admin.PulsarAdminException.wrap(PulsarAdminException.java:252)
  	at org.apache.pulsar.client.admin.internal.BaseResource.sync(BaseResource.java:351)
  	at org.apache.pulsar.client.admin.internal.NamespacesImpl.deleteNamespace(NamespacesImpl.java:231)
  	at org.apache.pulsar.broker.service.BrokerTestBase.deleteNamespaceGraceFully(BrokerTestBase.java:156)
  	at org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.deleteNamespaceGraceFully(MockedPulsarServiceBaseTest.java:673)
  	at org.apache.pulsar.broker.admin.AdminApi2Test.resetClusters(AdminApi2Test.java:167)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:61)
  	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:366)
  	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:320)
  	at org.testng.internal.TestInvoker.runConfigMethods(TestInvoker.java:701)
  	at org.testng.internal.TestInvoker.runAfterGroupsConfigurations(TestInvoker.java:677)
  	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:661)
  	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
  	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
  	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
  	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
  	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
  	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
  	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
  	at org.testng.TestRunner.privateRun(TestRunner.java:764)
  	at org.testng.TestRunner.run(TestRunner.java:585)
  	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
  	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
  	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
  	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
  	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
  	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
  	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
  	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
  	at org.testng.TestNG.runSuites(TestNG.java:1069)
  	at org.testng.TestNG.run(TestNG.java:1037)
  	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
  	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
  	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeLazy(TestNGDirectoryTestSuite.java:123)
  	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:90)
  	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
  	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
  	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
  	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
  	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
  	Suppressed: org.apache.pulsar.client.admin.PulsarAdminException$ServerSideErrorException: 
   --- An unexpected error occurred in the server ---
  
  Message: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  
  Stacktrace:
  
  org.apache.pulsar.metadata.api.MetadataStoreException: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.getException(ZKMetadataStore.java:449)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.handleDeleteResult(ZKMetadataStore.java:279)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.lambda$batchOperation$7(ZKMetadataStore.java:191)
  	at org.apache.zookeeper.MockZooKeeper.multi(MockZooKeeper.java:1006)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.batchOperation(ZKMetadataStore.java:165)
  	at org.apache.pulsar.metadata.impl.batching.AbstractBatchedMetadataStore.flush(AbstractBatchedMetadataStore.java:117)
  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
  	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
  	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  	at java.base/java.lang.Thread.run(Thread.java:833)
  Caused by: org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  	at org.apache.zookeeper.KeeperException.create(KeeperException.java:132)
  	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
  	at org.apache.pulsar.metadata.impl.ZKMetadataStore.getException(ZKMetadataStore.java:439)
  	... 12 more
  
  		at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:272)
  		at org.apache.pulsar.client.admin.internal.BaseResource$4.failed(BaseResource.java:237)
  		at org.glassfish.jersey.client.JerseyInvocation$1.failed(JerseyInvocation.java:882)
  		at org.glassfish.jersey.client.JerseyInvocation$1.completed(JerseyInvocation.java:863)
  		at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:229)
  		at org.glassfish.jersey.client.ClientRuntime.access$200(ClientRuntime.java:62)
  		at org.glassfish.jersey.client.ClientRuntime$2.lambda$response$0(ClientRuntime.java:173)
  		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
  		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
  		at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
  		at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
  		at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
  		at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:288)
  		at org.glassfish.jersey.client.ClientRuntime$2.response(ClientRuntime.java:173)
  		at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$apply$1(AsyncHttpConnector.java:251)
  		at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
  		at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
  		at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
  		at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
  		at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$retryOperation$4(AsyncHttpConnector.java:293)
  		at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
  		at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
  		at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
  		at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
  		at org.asynchttpclient.netty.NettyResponseFuture.loadContent(NettyResponseFuture.java:222)
  		at org.asynchttpclient.netty.NettyResponseFuture.done(NettyResponseFuture.java:257)
  		at org.asynchttpclient.netty.handler.AsyncHttpClientHandler.finishUpdate(AsyncHttpClientHandler.java:241)
  		at org.asynchttpclient.netty.handler.HttpHandler.handleChunk(HttpHandler.java:114)
  		at org.asynchttpclient.netty.handler.HttpHandler.handleRead(HttpHandler.java:143)
  		at org.asynchttpclient.netty.handler.AsyncHttpClientHandler.channelRead(AsyncHttpClientHandler.java:78)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
  		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
  		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
  		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
  		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
  		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
  		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
  		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
  		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
  		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
  		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
  		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
  		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
  		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
  		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
  		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
  		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
  		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
  		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
  		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  		at java.base/java.lang.Thread.run(Thread.java:833)
  	Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent
  		at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:960)
  		at org.glassfish.jersey.client.JerseyInvocation.access$700(JerseyInvocation.java:82)
  		... 54 more
  Caused by: [CIRCULAR REFERENCE: javax.ws.rs.InternalServerErrorException: HTTP 500 org.apache.zookeeper.KeeperException$NotEmptyException: KeeperErrorCode = Directory not empty for /managed-ledgers/prop-xyz/ns1/persistent]
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java
@@ -164,7 +164,8 @@ public class AdminApi2Test extends MockedPulsarServiceBaseTest {
         pulsar.getConfiguration().setForceDeleteNamespaceAllowed(true);
         for (String tenant : admin.tenants().getTenants()) {
             for (String namespace : admin.namespaces().getNamespaces(tenant)) {
-                deleteNamespaceGraceFully(namespace, true);
+                deleteNamespaceGraceFullyByMultiPulsars(namespace, true, admin, pulsar,
+                        mockPulsarSetup.getPulsar());
             }
             admin.tenants().deleteTenant(tenant, true);
         }

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java
@@ -668,19 +668,27 @@ public abstract class MockedPulsarServiceBaseTest extends TestRetrySupport {
     }
 
     /**
-     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarService, PulsarAdmin)}
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
      */
     protected void deleteNamespaceGraceFully(String ns, boolean force)
             throws Exception {
-        BrokerTestBase.deleteNamespaceGraceFully(ns, force, pulsar, admin);
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsar);
     }
 
     /**
-     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarService, PulsarAdmin)}
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
      */
     protected void deleteNamespaceGraceFully(String ns, boolean force, PulsarAdmin admin)
             throws Exception {
-        BrokerTestBase.deleteNamespaceGraceFully(ns, force, pulsar, admin);
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsar);
+    }
+
+    /**
+     * see see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    protected void deleteNamespaceGraceFullyByMultiPulsars(String ns, boolean force, PulsarAdmin admin,
+                                                           PulsarService...pulsars) throws Exception {
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsars);
     }
 
     private static final Logger log = LoggerFactory.getLogger(MockedPulsarServiceBaseTest.class);

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/auth/MockedPulsarServiceBaseTest.java
@@ -63,6 +63,7 @@ import org.apache.pulsar.broker.ServiceConfiguration;
 import org.apache.pulsar.broker.intercept.CounterBrokerInterceptor;
 import org.apache.pulsar.broker.namespace.NamespaceService;
 import org.apache.pulsar.broker.service.BrokerTestBase;
+import org.apache.pulsar.broker.service.CanPausedNamespaceService;
 import org.apache.pulsar.broker.service.PulsarMetadataEventSynchronizer;
 import org.apache.pulsar.client.admin.PulsarAdmin;
 import org.apache.pulsar.client.admin.PulsarAdminBuilder;
@@ -381,7 +382,8 @@ public abstract class MockedPulsarServiceBaseTest extends TestRetrySupport {
         doReturn(configSynchronizer != null ? createConfigurationMetadataStore(configSynchronizer)
                 : createConfigurationMetadataStore()).when(pulsar).createConfigurationMetadataStore(any());
 
-        Supplier<NamespaceService> namespaceServiceSupplier = () -> spyWithClassAndConstructorArgs(NamespaceService.class, pulsar);
+        Supplier<NamespaceService> namespaceServiceSupplier =
+                () -> spyWithClassAndConstructorArgs(CanPausedNamespaceService.class, pulsar);
         doReturn(namespaceServiceSupplier).when(pulsar).getNamespaceServiceProvider();
 
         doReturn(sameThreadOrderedSafeExecutor).when(pulsar).getOrderedExecutor();

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/OwnerShipForCurrentServerTestBase.java
@@ -45,6 +45,7 @@ import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.ServiceConfiguration;
 import org.apache.pulsar.broker.auth.SameThreadOrderedSafeExecutor;
 import org.apache.pulsar.broker.intercept.CounterBrokerInterceptor;
+import org.apache.pulsar.broker.service.CanPausedNamespaceService;
 import org.apache.pulsar.client.admin.PulsarAdmin;
 import org.apache.pulsar.client.api.PulsarClient;
 import org.apache.pulsar.metadata.api.extended.MetadataStoreExtended;
@@ -131,7 +132,8 @@ public class OwnerShipForCurrentServerTestBase {
         MockZooKeeperSession mockZooKeeperSession = MockZooKeeperSession.newInstance(mockZooKeeper);
         doReturn(new ZKMetadataStore(mockZooKeeperSession)).when(pulsar).createLocalMetadataStore(null);
         doReturn(new ZKMetadataStore(mockZooKeeperSession)).when(pulsar).createConfigurationMetadataStore(null);
-        Supplier<NamespaceService> namespaceServiceSupplier = () -> spyWithClassAndConstructorArgs(NamespaceService.class, pulsar);
+        Supplier<NamespaceService> namespaceServiceSupplier = () -> spyWithClassAndConstructorArgs(
+                CanPausedNamespaceService.class, pulsar);
         doReturn(namespaceServiceSupplier).when(pulsar).getNamespaceServiceProvider();
 
         SameThreadOrderedSafeExecutor executor = new SameThreadOrderedSafeExecutor();

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BacklogQuotaManagerTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BacklogQuotaManagerTest.java
@@ -26,6 +26,7 @@ import com.google.common.collect.Sets;
 import java.net.URL;
 import java.time.Duration;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Optional;
@@ -81,11 +82,27 @@ public class BacklogQuotaManagerTest {
     private static final int MAX_ENTRIES_PER_LEDGER = 5;
 
     /**
-     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarService, PulsarAdmin)}
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
      */
     protected void deleteNamespaceGraceFully(String ns, boolean force)
             throws Exception {
-        BrokerTestBase.deleteNamespaceGraceFully(ns, force, pulsar, admin);
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsar);
+    }
+
+    /**
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    protected void deleteNamespaceGraceFully(String ns, boolean force, PulsarAdmin admin)
+            throws Exception {
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsar);
+    }
+
+    /**
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    protected void deleteNamespaceGraceFullyByMultiPulsars(String ns, boolean force, PulsarAdmin admin,
+                                                           PulsarService...pulsars) throws Exception {
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsars);
     }
 
     @DataProvider(name = "backlogQuotaSizeGB")

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
@@ -23,6 +23,8 @@ import static org.apache.pulsar.compaction.Compactor.COMPACTION_SUBSCRIPTION;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -133,62 +135,88 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
     }
 
     /**
-     * see {@link #deleteNamespaceGraceFully}
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
      */
     protected void deleteNamespaceGraceFully(String ns, boolean force)
             throws Exception {
-        deleteNamespaceGraceFully(ns, force, pulsar, admin);
+        deleteNamespaceGraceFully(ns, force, admin, pulsar);
+    }
+
+    /**
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    public static void deleteNamespaceGraceFully(String ns, boolean force, PulsarAdmin admin, PulsarService...pulsars)
+            throws Exception {
+        deleteNamespaceGraceFully(ns, force, admin, Arrays.asList(pulsars));
     }
 
     /**
      * Wait until system topic "__change_event" and subscription "__compaction" are created, and then delete the namespace.
      */
-    public static void deleteNamespaceGraceFully(String ns, boolean force, PulsarService pulsar, PulsarAdmin admin)
-            throws Exception {
+    public static void deleteNamespaceGraceFully(String ns, boolean force, PulsarAdmin admin,
+                                                 Collection<PulsarService> pulsars) throws Exception {
         // namespace v1 should not wait system topic create.
         if (ns.split("/").length > 2){
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
+
         // If disabled system-topic, should not wait system topic create.
-        if (!pulsar.getConfiguration().isSystemTopicEnabled()){
-            admin.namespaces().deleteNamespace(ns, force);
-            return;
+        boolean allBrokerDisabledSystemTopic = true;
+        for (PulsarService pulsar : pulsars) {
+            if (!pulsar.getConfiguration().isSystemTopicEnabled()) {
+                continue;
+            }
+            TopicPoliciesService topicPoliciesService = pulsar.getTopicPoliciesService();
+            if (!(topicPoliciesService instanceof SystemTopicBasedTopicPoliciesService)) {
+                continue;
+            }
+            allBrokerDisabledSystemTopic = false;
         }
-        TopicPoliciesService topicPoliciesService = pulsar.getTopicPoliciesService();
-        if (!(topicPoliciesService instanceof SystemTopicBasedTopicPoliciesService)){
+        if (allBrokerDisabledSystemTopic){
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
 
-        // Prevents new events from triggering system topic creation.
-        CanPausedNamespaceService canPausedNamespaceService = (CanPausedNamespaceService) pulsar.getNamespaceService();
-        canPausedNamespaceService.pause();
+        // Stop trigger "onNamespaceBundleOwned".
+        List<CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> createReaderTasks = new ArrayList<>();
+        List<String> lockedBundles = new ArrayList<>();
+        for (PulsarService pulsar : pulsars) {
+            // Prevents new events from triggering system topic creation.
+            CanPausedNamespaceService canPausedNamespaceService = (CanPausedNamespaceService) pulsar.getNamespaceService();
+            canPausedNamespaceService.pause();
+            // Determines whether the creation of System topic is triggered.
+            // If readerCaches contains namespace, the creation of System topic already triggered.
+            SystemTopicBasedTopicPoliciesService systemTopicBasedTopicPoliciesService =
+                    (SystemTopicBasedTopicPoliciesService) pulsar.getTopicPoliciesService();
+            Map<NamespaceName, CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> readerCaches =
+                    WhiteboxImpl.getInternalState(systemTopicBasedTopicPoliciesService, "readerCaches");
+            if (readerCaches.containsKey(NamespaceName.get(ns))) {
+                createReaderTasks.add(readerCaches.get(NamespaceName.get(ns)));
+            }
 
-        // Determines whether the creation of System topic is triggered.
-        // If readerCaches contains namespace, the creation of System topic already triggered.
-        SystemTopicBasedTopicPoliciesService systemTopicBasedTopicPoliciesService =
-                (SystemTopicBasedTopicPoliciesService) pulsar.getTopicPoliciesService();
-        Map<NamespaceName, CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> readerCaches =
-                WhiteboxImpl.getInternalState(systemTopicBasedTopicPoliciesService, "readerCaches");
-        // If no bundle has been loaded, then the System Topic will not trigger creation.
-        LockManager lockManager = pulsar.getCoordinationService().getLockManager(NamespaceEphemeralData.class);
-        List<String> lockedBundles = (List<String>) lockManager.listLocks("/namespace" + "/" + ns).join();
-        FutureUtil.waitForAll(readerCaches.values()).join();
+            // If no bundle has been loaded, then the System Topic will not trigger creation.
+            LockManager lockManager = pulsar.getCoordinationService().getLockManager(NamespaceEphemeralData.class);
+            lockedBundles.addAll((List<String>) lockManager.listLocks("/namespace" + "/" + ns).join());
+        }
+        // Wait all reader-create tasks.
+        FutureUtil.waitForAll(createReaderTasks).join();
 
         // If the bundle elect has not yet been triggered, skip wait.
-        if (CollectionUtils.isEmpty(lockedBundles) && !readerCaches.containsKey(NamespaceName.get(ns))){
+        if (CollectionUtils.isEmpty(lockedBundles) && createReaderTasks.isEmpty()){
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
+
         // Trigger change event topic create.
+        PulsarService firstPulsar = pulsars.iterator().next();
         NamespaceName namespace = NamespaceName.get(ns);
         NamespaceBundle namespaceBundle = mock(NamespaceBundle.class);
         when(namespaceBundle.getNamespaceObject()).thenReturn(namespace);
-        pulsar.getTopicPoliciesService().addOwnedNamespaceBundleAsync(namespaceBundle);
+        firstPulsar.getTopicPoliciesService().addOwnedNamespaceBundleAsync(namespaceBundle);
         // Wait for change event topic and compaction create finish.
-        String allowAutoTopicCreationType = pulsar.getConfiguration().getAllowAutoTopicCreationType();
-        int defaultNumPartitions = pulsar.getConfiguration().getDefaultNumPartitions();
+        String allowAutoTopicCreationType = firstPulsar.getConfiguration().getAllowAutoTopicCreationType();
+        int defaultNumPartitions = firstPulsar.getConfiguration().getDefaultNumPartitions();
         ArrayList<String> expectChangeEventTopics = new ArrayList<>();
         if ("non-partitioned".equals(allowAutoTopicCreationType)){
             String t = String.format("persistent://%s/%s", ns, NAMESPACE_EVENTS_LOCAL_NAME);
@@ -202,7 +230,7 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
         Awaitility.await().until(() -> {
             boolean finished = true;
             for (String changeEventTopicName : expectChangeEventTopics){
-                boolean bundleExists = pulsar.getNamespaceService()
+                boolean bundleExists = firstPulsar.getNamespaceService()
                         .checkTopicOwnership(TopicName.get(changeEventTopicName))
                         .exceptionally(ex -> false).join();
                 if (!bundleExists){
@@ -210,7 +238,7 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
                     break;
                 }
                 CompletableFuture<Optional<Topic>> completableFuture =
-                        pulsar.getBrokerService().getTopic(changeEventTopicName, false);
+                        firstPulsar.getBrokerService().getTopic(changeEventTopicName, false);
                 if (completableFuture == null){
                     finished = false;
                     break;
@@ -231,6 +259,8 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
             }
             return finished;
         });
+
+        // Do delete.
         int retryTimes = 3;
         while (true) {
             try {
@@ -248,7 +278,14 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
                 throw ex;
             }
         }
-        canPausedNamespaceService.resume();
+
+        // Resume trigger "onNamespaceBundleOwned".
+        for (PulsarService pulsarService : pulsars) {
+            // Prevents new events from triggering system topic creation.
+            CanPausedNamespaceService canPausedNamespaceService =
+                    (CanPausedNamespaceService) pulsarService.getNamespaceService();
+            canPausedNamespaceService.resume();
+        }
     }
 
     private static final Logger LOG = LoggerFactory.getLogger(BrokerTestBase.class);

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
@@ -24,6 +24,7 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import org.apache.commons.collections4.CollectionUtils;
@@ -32,8 +33,10 @@ import org.apache.pulsar.broker.ServiceConfiguration;
 import org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest;
 import org.apache.pulsar.broker.namespace.NamespaceEphemeralData;
 import org.apache.pulsar.broker.service.persistent.PersistentTopic;
+import org.apache.pulsar.broker.systopic.SystemTopicClient;
 import org.apache.pulsar.client.admin.PulsarAdmin;
 import org.apache.pulsar.client.admin.PulsarAdminException;
+import org.apache.pulsar.common.events.PulsarEvent;
 import org.apache.pulsar.common.naming.NamespaceBundle;
 import org.apache.pulsar.common.naming.NamespaceName;
 import org.apache.pulsar.common.naming.SystemTopicNames;
@@ -41,9 +44,11 @@ import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.partition.PartitionedTopicMetadata;
 import org.apache.pulsar.common.policies.data.ClusterData;
 import org.apache.pulsar.common.policies.data.TenantInfoImpl;
+import org.apache.pulsar.common.util.FutureUtil;
 import org.apache.pulsar.metadata.api.MetadataStoreException;
 import org.apache.pulsar.metadata.api.coordination.LockManager;
 import org.awaitility.Awaitility;
+import org.awaitility.reflect.WhiteboxImpl;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -145,14 +150,34 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
+        // If disabled system-topic, should not wait system topic create.
         if (!pulsar.getConfiguration().isSystemTopicEnabled()){
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
+        TopicPoliciesService topicPoliciesService = pulsar.getTopicPoliciesService();
+        if (!(topicPoliciesService instanceof SystemTopicBasedTopicPoliciesService)){
+            admin.namespaces().deleteNamespace(ns, force);
+            return;
+        }
+
+        // Prevents new events from triggering system topic creation.
+        CanPausedNamespaceService canPausedNamespaceService = (CanPausedNamespaceService) pulsar.getNamespaceService();
+        canPausedNamespaceService.pause();
+
+        // Determines whether the creation of System topic is triggered.
+        // If readerCaches contains namespace, the creation of System topic already triggered.
+        SystemTopicBasedTopicPoliciesService systemTopicBasedTopicPoliciesService =
+                (SystemTopicBasedTopicPoliciesService) pulsar.getTopicPoliciesService();
+        Map<NamespaceName, CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> readerCaches =
+                WhiteboxImpl.getInternalState(systemTopicBasedTopicPoliciesService, "readerCaches");
         // If no bundle has been loaded, then the System Topic will not trigger creation.
         LockManager lockManager = pulsar.getCoordinationService().getLockManager(NamespaceEphemeralData.class);
         List<String> lockedBundles = (List<String>) lockManager.listLocks("/namespace" + "/" + ns).join();
-        if (CollectionUtils.isEmpty(lockedBundles)){
+        FutureUtil.waitForAll(readerCaches.values()).join();
+
+        // If the bundle elect has not yet been triggered, skip wait.
+        if (CollectionUtils.isEmpty(lockedBundles) && !readerCaches.containsKey(NamespaceName.get(ns))){
             admin.namespaces().deleteNamespace(ns, force);
             return;
         }
@@ -223,6 +248,7 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
                 throw ex;
             }
         }
+        canPausedNamespaceService.resume();
     }
 
     private static final Logger LOG = LoggerFactory.getLogger(BrokerTestBase.class);

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerTestBase.java
@@ -185,19 +185,22 @@ public abstract class BrokerTestBase extends MockedPulsarServiceBaseTest {
             // Prevents new events from triggering system topic creation.
             CanPausedNamespaceService canPausedNamespaceService = (CanPausedNamespaceService) pulsar.getNamespaceService();
             canPausedNamespaceService.pause();
-            // Determines whether the creation of System topic is triggered.
-            // If readerCaches contains namespace, the creation of System topic already triggered.
-            SystemTopicBasedTopicPoliciesService systemTopicBasedTopicPoliciesService =
-                    (SystemTopicBasedTopicPoliciesService) pulsar.getTopicPoliciesService();
-            Map<NamespaceName, CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> readerCaches =
-                    WhiteboxImpl.getInternalState(systemTopicBasedTopicPoliciesService, "readerCaches");
-            if (readerCaches.containsKey(NamespaceName.get(ns))) {
-                createReaderTasks.add(readerCaches.get(NamespaceName.get(ns)));
-            }
 
             // If no bundle has been loaded, then the System Topic will not trigger creation.
             LockManager lockManager = pulsar.getCoordinationService().getLockManager(NamespaceEphemeralData.class);
             lockedBundles.addAll((List<String>) lockManager.listLocks("/namespace" + "/" + ns).join());
+
+            // Determines whether the creation of System topic is triggered.
+            // If readerCaches contains namespace, the creation of System topic already triggered.
+            TopicPoliciesService topicPoliciesService = pulsar.getTopicPoliciesService();
+            if (topicPoliciesService instanceof
+                    SystemTopicBasedTopicPoliciesService systemTopicBasedTopicPoliciesService) {
+                Map<NamespaceName, CompletableFuture<SystemTopicClient.Reader<PulsarEvent>>> readerCaches =
+                        WhiteboxImpl.getInternalState(systemTopicBasedTopicPoliciesService, "readerCaches");
+                if (readerCaches.containsKey(NamespaceName.get(ns))) {
+                    createReaderTasks.add(readerCaches.get(NamespaceName.get(ns)));
+                }
+            }
         }
         // Wait all reader-create tasks.
         FutureUtil.waitForAll(createReaderTasks).join();

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/CanPausedNamespaceService.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/CanPausedNamespaceService.java
@@ -1,0 +1,67 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.pulsar.broker.PulsarService;
+import org.apache.pulsar.broker.namespace.NamespaceService;
+import org.apache.pulsar.common.naming.NamespaceBundle;
+
+public class CanPausedNamespaceService extends NamespaceService {
+
+    private AtomicBoolean paused = new AtomicBoolean();
+
+    private ReentrantLock lock = new ReentrantLock();
+
+    public CanPausedNamespaceService(PulsarService pulsar) {
+        super(pulsar);
+    }
+
+    @Override
+    protected void onNamespaceBundleOwned(NamespaceBundle bundle) {
+        lock.lock();
+        try {
+            if (paused.get()){
+                return;
+            }
+            super.onNamespaceBundleOwned(bundle);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void pause(){
+        lock.lock();
+        try {
+            paused.set(true);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void resume(){
+        lock.lock();
+        try {
+            paused.set(false);
+        } finally {
+            lock.unlock();
+        }
+    }
+}

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/CanPausedNamespaceService.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/CanPausedNamespaceService.java
@@ -18,7 +18,6 @@
  */
 package org.apache.pulsar.broker.service;
 
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantLock;
 import org.apache.pulsar.broker.PulsarService;
 import org.apache.pulsar.broker.namespace.NamespaceService;
@@ -26,7 +25,7 @@ import org.apache.pulsar.common.naming.NamespaceBundle;
 
 public class CanPausedNamespaceService extends NamespaceService {
 
-    private AtomicBoolean paused = new AtomicBoolean();
+    private volatile boolean paused = false;
 
     private ReentrantLock lock = new ReentrantLock();
 
@@ -38,7 +37,7 @@ public class CanPausedNamespaceService extends NamespaceService {
     protected void onNamespaceBundleOwned(NamespaceBundle bundle) {
         lock.lock();
         try {
-            if (paused.get()){
+            if (paused){
                 return;
             }
             super.onNamespaceBundleOwned(bundle);
@@ -50,7 +49,7 @@ public class CanPausedNamespaceService extends NamespaceService {
     public void pause(){
         lock.lock();
         try {
-            paused.set(true);
+            paused = true;
         } finally {
             lock.unlock();
         }
@@ -59,7 +58,7 @@ public class CanPausedNamespaceService extends NamespaceService {
     public void resume(){
         lock.lock();
         try {
-            paused.set(false);
+            paused = false;
         } finally {
             lock.unlock();
         }

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTestBase.java
@@ -27,6 +27,7 @@ import com.google.common.util.concurrent.MoreExecutors;
 import io.netty.channel.EventLoopGroup;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -331,10 +332,26 @@ public abstract class TransactionTestBase extends TestRetrySupport {
     }
 
     /**
-     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarService, PulsarAdmin)}
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
      */
     protected void deleteNamespaceGraceFully(String ns, boolean force)
             throws Exception {
-        BrokerTestBase.deleteNamespaceGraceFully(ns, force, pulsarServiceList.get(0), admin);
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsarServiceList);
+    }
+
+    /**
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    protected void deleteNamespaceGraceFully(String ns, boolean force, PulsarAdmin admin)
+            throws Exception {
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsarServiceList);
+    }
+
+    /**
+     * see {@link BrokerTestBase#deleteNamespaceGraceFully(String, boolean, PulsarAdmin, Collection)}
+     */
+    protected void deleteNamespaceGraceFullyByMultiPulsars(String ns, boolean force, PulsarAdmin admin,
+                                                           Collection<PulsarService> pulsars) throws Exception {
+        BrokerTestBase.deleteNamespaceGraceFully(ns, force, admin, pulsars);
     }
 }

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTestBase.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTestBase.java
@@ -48,6 +48,7 @@ import org.apache.pulsar.broker.auth.SameThreadOrderedSafeExecutor;
 import org.apache.pulsar.broker.intercept.CounterBrokerInterceptor;
 import org.apache.pulsar.broker.namespace.NamespaceService;
 import org.apache.pulsar.broker.service.BrokerTestBase;
+import org.apache.pulsar.broker.service.CanPausedNamespaceService;
 import org.apache.pulsar.client.admin.PulsarAdmin;
 import org.apache.pulsar.client.api.PulsarClient;
 import org.apache.pulsar.common.naming.NamespaceName;
@@ -195,7 +196,8 @@ public abstract class TransactionTestBase extends TestRetrySupport {
         MockZooKeeperSession mockZooKeeperSession = MockZooKeeperSession.newInstance(mockZooKeeper);
         doReturn(new ZKMetadataStore(mockZooKeeperSession)).when(pulsar).createLocalMetadataStore(null);
         doReturn(new ZKMetadataStore(mockZooKeeperSession)).when(pulsar).createConfigurationMetadataStore(null);
-        Supplier<NamespaceService> namespaceServiceSupplier = () -> spyWithClassAndConstructorArgs(NamespaceService.class, pulsar);
+        Supplier<NamespaceService> namespaceServiceSupplier =
+                () -> spyWithClassAndConstructorArgs(CanPausedNamespaceService.class, pulsar);
         doReturn(namespaceServiceSupplier).when(pulsar).getNamespaceServiceProvider();
 
         SameThreadOrderedSafeExecutor executor = new SameThreadOrderedSafeExecutor();

--- a/pulsar-broker/src/test/java/org/apache/pulsar/client/api/DispatcherBlockConsumerTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/client/api/DispatcherBlockConsumerTest.java
@@ -50,6 +50,7 @@ import java.util.stream.Collectors;
 import lombok.Cleanup;
 import org.apache.pulsar.broker.namespace.NamespaceService;
 import org.apache.pulsar.broker.service.BrokerService;
+import org.apache.pulsar.broker.service.CanPausedNamespaceService;
 import org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers;
 import org.apache.pulsar.broker.service.persistent.PersistentTopic;
 import org.apache.pulsar.client.impl.ConsumerImpl;
@@ -633,7 +634,8 @@ public class DispatcherBlockConsumerTest extends ProducerConsumerBase {
         // if broker unload bundle gracefully then cursor metadata recovered from zk else from ledger
         if (unloadBundleGracefully) {
             // set clean namespace which will not let broker unload bundle gracefully: stop broker
-            Supplier<NamespaceService> namespaceServiceSupplier = () -> spyWithClassAndConstructorArgs(NamespaceService.class, pulsar);
+            Supplier<NamespaceService> namespaceServiceSupplier =
+                    () -> spyWithClassAndConstructorArgs(CanPausedNamespaceService.class, pulsar);
             doReturn(namespaceServiceSupplier).when(pulsar).getNamespaceServiceProvider();
         }
         stopBroker();

